README.md
# kerrgeodesic_gw

A [SageMath](https://www.sagemath.org/) package to compute gravitational radiation from material orbiting a Kerr black hole

This package makes use of SageMath functionalities developed through the [SageManifolds](https://sagemanifolds.obspm.fr/) project and is part of the [Black Hole Perturbation Toolkit](https://bhptoolkit.org/).

## Installation

### Requirements

This package requires the Python-based free mathematics software system [SageMath](https://www.sagemath.org/) (at least version 9.3).
For versions of SageMath lower than 9.3, only version 0.3.2 of the package will work.

*NB:* the version of SageMath shipped with Ubuntu 20.04 is only 9.0; instead of
the Ubuntu package `sagemath`, install then the most recent binary for Ubuntu 20.04
from [SageMath download page](https://www.sagemath.org/download-linux.html).

### Simple installation from PyPI

It suffices to run

    sage -pip install kerrgeodesic_gw

to have the package ready to use in SageMath.
See however *install from source* below if you want to build a
local version of the documentation or modify the source files (development).

*NB1 (Ubuntu users):* if you have installed SageMath as the system package `sagemath` (e.g. via `apt-get install`), the installation is system wide and does not allow to use `-pip install` at the user level. Install instead SageMath from the most recent binary tar ball for Ubuntu
from [SageMath download page](https://www.sagemath.org/download-linux.html).

*NB2:* on the [CoCalc](https://cocalc.com) cloud computing platform, you need
to add the option `--user`, i.e. open a terminal and run

    sage -pip install --user kerrgeodesic_gw


Here is the [kerrgeodesic_gw page](https://pypi.org/project/kerrgeodesic-gw/) on PyPI (the Python Package Index).

### Install from source

Download the source from the git repository:

    git clone https://github.com/BlackHolePerturbationToolkit/kerrgeodesic_gw.git

This creates a directory `kerrgeodesic_gw`.

Run

    sage -pip install --upgrade --no-index -v kerrgeodesic_gw

to install the package in SageMath.
A shortcut of the above command is provided by the `Makefile` distributed with the package:

    cd kerrgeodesic_gw
    make install

*NB:* on [CoCalc](https://cocalc.com), you need to add the option `--user`, i.e. open a terminal and run

    git clone https://github.com/BlackHolePerturbationToolkit/kerrgeodesic_gw.git
    sage -pip install --user --upgrade --no-index -v kerrgeodesic_gw

#### Install for development

If you plan to edit the package source, you should add the option `-e` to the pip install, i.e. run

    sage -pip install --upgrade --no-index -v -e kerrgeodesic_gw

or equivalently

    cd kerrgeodesic_gw
    make develop

## Usage

Once the package is installed, you can use it in SageMath, like for instance:

    sage: from kerrgeodesic_gw import spin_weighted_spherical_harmonic
    sage: theta, phi = var('theta phi')
    sage: spin_weighted_spherical_harmonic(-2, 2, 1, theta, phi)
    1/4*(sqrt(5)*cos(theta) + sqrt(5))*e^(I*phi)*sin(theta)/sqrt(pi)


## Tests

This package is configured for tests written in the documentation strings of the source files, also known as *doctests*.
You may then test the install by running, from the root of the package tree
(i.e. the directory kerrgeodesic_gw created by the `git clone`),

    sage -t kerrgeodesic_gw

You should then get the message `All tests passed!`

Alternatively, you can run (from the same directory)

    make test


## Documentation

The package documentation can be generated using SageMath's [Sphinx](https://www.sphinx-doc.org/) installation:

    cd docs
    sage -sh -c "make html"

A shorthand of the above is

    make doc

The html reference manual is then at

    kerrgeodesic_gw/docs/build/html/index.html

For the LaTeX documentation, use

    make doc-pdf

The pdf reference manual is then

    kerrgeodesic_gw/docs/build/latex/kerrgeodesic_gw.pdf

### Online documentation

- [Reference manual](https://sagemanifolds.obspm.fr/kerrgeodesic_gw/reference/)
  ([PDF](https://sagemanifolds.obspm.fr/kerrgeodesic_gw/kerrgeodesic_gw.pdf))
- [Article describing the formulas implemented in the package](https://doi.org/10.1051/0004-6361/201935406) *(open access)*
- Demo notebooks:

  - [Spin-weighted spheroidal harmonics](https://nbviewer.jupyter.org/github/BlackHolePerturbationToolkit/kerrgeodesic_gw/blob/master/Notebooks/basic_kerrgeodesic_gw.ipynb) [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/BlackHolePerturbationToolkit/kerrgeodesic_gw/master?filepath=Notebooks/basic_kerrgeodesic_gw.ipynb)
  - [Timelike and null geodesics in Kerr spacetime](https://nbviewer.jupyter.org/github/BlackHolePerturbationToolkit/kerrgeodesic_gw/blob/master/Notebooks/Kerr_geodesics.ipynb) [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/BlackHolePerturbationToolkit/kerrgeodesic_gw/master?filepath=Notebooks/Kerr_geodesics.ipynb)
  - [Gravitational waves from circular orbits around a Kerr black hole](https://nbviewer.jupyter.org/github/BlackHolePerturbationToolkit/kerrgeodesic_gw/blob/master/Notebooks/grav_waves_circular.ipynb) [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/BlackHolePerturbationToolkit/kerrgeodesic_gw/master?filepath=Notebooks/grav_waves_circular.ipynb)
  - [More on gravitational waves from circular orbits](https://nbviewer.jupyter.org/github/BlackHolePerturbationToolkit/kerrgeodesic_gw/blob/master/Notebooks/gw_single_particle.ipynb) [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/BlackHolePerturbationToolkit/kerrgeodesic_gw/master?filepath=Notebooks/gw_single_particle.ipynb)

- For the tensor calculus functionalities of the
  [KerrBH](https://sagemanifolds.obspm.fr/kerrgeodesic_gw/reference/kerr_spacetime.html)
  class provided by the package, see these examples:
  [Kerr 1](https://nbviewer.jupyter.org/github/sagemanifolds/SageManifolds/blob/master/Notebooks/SM_Kerr.ipynb),
  [Kerr 2](https://nbviewer.jupyter.org/github/sagemanifolds/SageManifolds/blob/master/Notebooks/SM_Kerr_Killing_tensor.ipynb),
  [Kerr 3](https://nbviewer.jupyter.org/github/sagemanifolds/SageManifolds/blob/master/Worksheets/v1.3/SM_Simon-Mars_Kerr.ipynb),
  and more generally [SageManifolds documentation](https://sagemanifolds.obspm.fr/documentation.html).


## Authors

- Eric Gourgoulhon
- Alexandre Le Tiec
- Frederic Vincent
- Niels Warburton

**Reference:** E. Gourgoulhon, A. Le Tiec, F. H. Vincent & N. Warburton: *Gravitational waves from bodies orbiting the Galactic center black hole and their detectability by LISA*, [A&A **627**, A92 (2019)](https://doi.org/10.1051/0004-6361/201935406) (preprint: [arXiv:1903.02049](https://arxiv.org/abs/1903.02049))


CHANGELOG.md
## kerrgeodesic_gw changelog

- v0.1   (7 Feb 2019): first PyPi release
- v0.2   (19 May 2019): add `utilities.graphics_inset`
- v0.3   (25 May 2020): add computation of generic geodesics
- v0.3.1 (2 Jun 2020): improve treatment of axes labels in `KerrGeodesic.plot`
- v0.3.2 (30 Dec 2020): update to SageMath 9.2 (SciPy 1.5.2) + new method `KerrGeodesic.evaluate_mu2`
- v0.4   (11 Jul 2021): update to SageMath 9.3
- v0.4.1 (14 Sep 2021): update to SageMath 9.4 (SciPy 1.6.3)
- v0.4.2 (15 Sep 2021): new location for the online reference manual


kerrgeodesic_gw/zinf.py
r"""
Functions `Z^\infty_{\ell m}(r)` giving the amplitude of the gravitational
radiation emitted by a particle on a circular orbit at radius `r` about
a Kerr black hole.

REFERENCES:

- \S. A. Teukolsky, Astrophys. J. **185**, 635 (1973)
- \S. Detweiler, Astrophys. J. **225**, 687 (1978)
- \M. Shibata, Phys. Rev. D **50**, 6297 (1994)
- \D. Kennefick, Phys. Rev. D **58**, 064012 (1998)
- \S. A. Hughes, Phys. Rev. D **61**, 084004 (2000) [:doi:`10.1103/PhysRevD.61.084004`]
- \E. Gourgoulhon, A. Le Tiec, F. Vincent, N. Warburton, Astron. Astrophys.
  **627**, A92 (2019) [:doi:`10.1051/0004-6361/201935406`]; :arXiv:`1903.02049`

"""
#******************************************************************************
#       Copyright (C) 2018 Eric Gourgoulhon <eric.gourgoulhon@obspm.fr>
#       Copyright (C) 2018 Niels Warburton <niels.warburton@ucd.ie>
#
#  Distributed under the terms of the GNU General Public License (GPL)
#  as published by the Free Software Foundation; either version 2 of
#  the License, or (at your option) any later version.
#                  http://www.gnu.org/licenses/
#******************************************************************************

import os
from sage.calculus.interpolation import spline
from sage.functions.other import sqrt
from sage.functions.log import ln
from sage.rings.real_double import RDF
from sage.rings.complex_double import CDF
from sage.symbolic.constants import pi, I

_cached_splines = {}

def Zinf_Schwarzchild_PN(l, m, r):
    r"""
    Amplitude factor of the mode `(\ell,m)` for a Schwarzschild BH at the 1.5PN
    level.

    The 1.5PN formulas are taken from E. Poisson, Phys. Rev. D **47**, 1497
    (1993), :doi:`10.1103/PhysRevD.47.1497`.

    INPUT:

    - ``l`` -- integer >= 2; the harmonic degree `\ell`
    - ``m`` -- integer within the range ``[-l, l]``; the azimuthal number `m`
    - ``r`` -- areal radius of the orbit (in units of `M`, the BH mass)

    OUTPUT:

    - coefficient `Z^\infty_{\ell m}(r)` (in units of `M^{-2}`)

    EXAMPLES::

        sage: from kerrgeodesic_gw import Zinf_Schwarzchild_PN
        sage: Zinf_Schwarzchild_PN(2, 2, 6.)  # tol 1.0e-13
        -0.00981450418730346 + 0.003855681972781947*I
        sage: Zinf_Schwarzchild_PN(5, 3, 6.)  # tol 1.0e-13
        -6.958527913913504e-05*I

    """
    if m < 0:
        return (-1)**l * Zinf_Schwarzchild_PN(l, -m, r).conjugate()
    v = 1./sqrt(r)
    if l == 2:
        b = RDF(sqrt(pi/5.)/r**4)
        if m == 1:
            return CDF(4./3.*I*b*v*(1 - 17./28.*v**2))
        if m == 2:
            return CDF(-16*b*(1 - 107./42.*v**2
                        + (2*pi + 4*I*(3*ln(2*v) - 0.839451001765134))*v**3))
    if l == 3:
        b = RDF(sqrt(pi/7.)/r**(4.5))
        if m == 1:
            return CDF(I/3.*b/sqrt(10.)*(1 - 8./3.*v**2))
        if m == 2:
            return CDF(-16./3.*b*v)
        if m == 3:
            return CDF(-81*I*b/sqrt(8.)*(1 - 4*v**2))
    if l == 4:
        b = RDF(sqrt(pi)/r**5)
        if m == 1:
            return CDF(I/105.*b/sqrt(2)*v)
        if m == 2:
            return CDF(-16./63.*b)
        if m == 3:
            return CDF(-81./5.*I*b/sqrt(14.)*v)
        if m == 4:
            return CDF(512./9.*b/sqrt(7.))
    if l == 5:
        b = RDF(sqrt(pi)/r**(5.5))
        if m == 1:
            return CDF(I/360.*b/sqrt(77.))
        if m == 2:
            return CDF(0)
        if m == 3:
            return CDF(-81./40.*I*b*sqrt(3./22.))
        if m == 4:
            return CDF(0)
        if m == 5:
            return CDF(3125./24.*I*b*sqrt(5./66.))
    raise NotImplementedError("{} not implemented".format((l, m)))

def Zinf(a, l, m, r, algorithm='spline'):
    r"""
    Amplitude factor of the mode `(\ell,m)`.

    The factor `Z^\infty_{\ell m}(r)` is obtained by spline interpolation
    of tabulated numerical solutions of the radial component of the
    Teukolsky equation.

    INPUT:

    - ``a`` -- BH angular momentum parameter (in units of `M`, the BH mass)
    - ``l`` -- integer >= 2; the harmonic degree `\ell`
    - ``m`` -- integer within the range ``[-l, l]``; the azimuthal number `m`
    - ``r`` -- areal radius of the orbit (in units of `M`)
    - ``algorithm`` -- (default: ``'spline'``) string describing the
      computational method; allowed values are

      - ``'spline'``: spline interpolation of tabulated data
      - ``'1.5PN'`` (only for ``a=0``): 1.5-post-Newtonian expansion following
        E. Poisson, Phys. Rev. D **47**, 1497 (1993)
        [:doi:`10.1103/PhysRevD.47.1497`], with a minus one factor
        accounting for a different convention for the metric signature.

    OUTPUT:

    - coefficient `Z^\infty_{\ell m}(r)` (in units of `M^{-2}`)

    EXAMPLES::

        sage: from kerrgeodesic_gw import Zinf
        sage: Zinf(0.98, 2, 2, 1.7)  # tol 1.0e-13
        -0.04302234478778856 + 0.28535368610053824*I
        sage: Zinf(0., 2, 2, 10.)  # tol 1.0e-13
        0.0011206407919254163 - 0.0003057608384581628*I
        sage: Zinf(0., 2, 2, 10., algorithm='1.5PN')  # tol 1.0e-13
        0.0011971529546749354 - 0.0003551610880408921*I

    """
    if m < 0:
        return (-1)**l * Zinf(a, l, -m, r, algorithm=algorithm).conjugate()
    if algorithm == '1.5PN':
        if a == 0:
            # the factor (-1) below accounts for a difference of signature with
            # Poisson (1993):
            return -Zinf_Schwarzchild_PN(l, m, r)
        raise ValueError("a must be zero for algorithm='1.5PN'")
    a = RDF(a)
    param = (a, l, m)
    if param in _cached_splines:
        splines = _cached_splines[param]
    else:
        file_name = "data/Zinf_a{:.1f}.dat".format(float(a)) if a <= 0.9 \
                    else "data/Zinf_a{:.2f}.dat".format(float(a))
        file_name = os.path.join(os.path.dirname(__file__), file_name)
        r_high_l = 20. if a <= 0.9 else 10.
        with open(file_name, "r") as data_file:
            lm_values = []        # l values up to 10 (for r <= r_high_l)
            lm_values_low = []    # l values up to 5 only (for r > r_high_l)
            for ld in range(2, 6):
                for md in range(1, ld+1):
                    lm_values_low.append((ld, md))
            for ld in range(2, 11):
                for md in range(1, ld+1):
                    lm_values.append((ld, md))
            Zreal = {}
            Zimag = {}
            for (ld, md) in lm_values:
                Zreal[(ld, md)] = []
                Zimag[(ld, md)] = []
            for line in data_file:
                items = line.split('\t')
                rd = RDF(items.pop(0))
                if rd <= r_high_l:
                    for (ld, md) in lm_values:
                        Zreal[(ld, md)].append((rd, RDF(items.pop(0))))
                        Zimag[(ld, md)].append((rd, RDF(items.pop(0))))
                else:
                    for (ld, md) in lm_values_low:
                        Zreal[(ld, md)].append((rd, RDF(items.pop(0))))
                        Zimag[(ld, md)].append((rd, RDF(items.pop(0))))
        for (ld, md) in lm_values:
            _cached_splines[(a,) + (ld, md)] = (spline(Zreal[(ld, md)]),
                                                spline(Zimag[(ld, md)]))
        if param not in _cached_splines:
            raise ValueError("Zinf: case (a, l, m) = {} not implemented".format(param))
        splines = _cached_splines[param]
    # The factor (-1)**(l+m) below accounts for a difference of convention
    # in the C++ code used to produce the data files
    return (-1)**(l+m)*CDF(splines[0](r), splines[1](r))

def _lmax(a, r):
    r"""
    Utility function giving the maximal value of the harmonic index `\ell`
    for tabulated values of `Z^\infty_{\ell m}(r)`.

    INPUT:

    - ``a`` -- BH angular momentum parameter (in units of `M`, the BH mass)
    - ``r`` -- areal radius of the orbit (in units of `M`)

    OUTPUT:

    - maximal value of `\ell` for which `Z^\infty_{\ell m}(r)` is tabulated

    EXAMPLES::

        sage: from kerrgeodesic_gw.zinf import _lmax
        sage: _lmax(0, 6.)
        10
        sage: _lmax(0, 40.)
        5
        sage: _lmax(0.9, 12.)
        10
        sage: _lmax(0.95, 12.)
        5

    """
    r_high_l = 20. if a <= 0.9 else 10.
    l_max = 10 if r <= r_high_l else 5
    return l_max
